pkg/planner/core/logical_plan_builder.go

```diff
@@ -670,7 +677,10 @@ func (b *PlanBuilder) buildJoin(ctx context.Context, joinNode *ast.Join) (base.L
 		}
 	} else if joinNode.On != nil {
 		b.curClause = onClause
+		originColPriv := b.checkColPriv
+		b.checkColPriv = reportColumnErrOption
```

Comment:
We need to explicitly enable the collection of columns here, for the `rewrite` in line 682.